File: bacongraph/saop.cpp
// a optimized SA implemented in c++
//
// same idea as 'sa.py',
// just randomly swap and a chosen one and a candidate one.
//
// but in each round, fastly calculate nearest bacon neighbors
// h*log(b) vs. h*b
//
// we utilize STL set's order feature (red-black tree actually underlying)
//
// one notable parameter that you might like to change is the scale
// how temperature decreses. for better results, you can use a higher scale
// variable (< 1), but the runtime is longer as well.
//
// g++ -O3 saop.cpp
// ./a.out < file  or  ./a.out file
#include <vector>
#include <set>
#include <algorithm>
#include <cstdio>
#include <cstdlib>
#include <ctime>
#include <cmath>

using namespace std;

typedef pair<int,int> point;

class findbacons {
	vector<point> houses, bacons, candidates;
	typedef set<pair<int, point> > set_t;
	vector<set_t> mindists;
public:
	findbacons(vector<point> &h, vector<point> &c, int b) : houses (h)
	{
		random_shuffle(c.begin(), c.end());
		for (vector<point>::iterator vi = c.begin(); vi != c.end(); ++vi) {
			if (vi < c.begin() + b)
				bacons.push_back(*vi);
			else
				candidates.push_back(*vi);
		}

		for (vector<point>::iterator vi = h.begin(); vi != h.end(); ++vi) {
			set_t s;
			for (int i = 0; i < b; i++)
				s.insert(make_pair(dist(*vi, bacons[i]), bacons[i]));
			mindists.push_back(s);
		}
	}

	inline int dist(point &a, point &b)
	{
		return abs(a.first - b.first) + abs(a.second - b.second);
	}

	inline double randpos()
	{
		return random() % RAND_MAX / (double) RAND_MAX;
	}

	inline int cost()
	{
		int ret = 0;
		for (vector<set_t>::iterator vi = mindists.begin();
		    vi != mindists.end(); ++vi)
			ret += vi->begin()->first;
		return ret;
	}

	inline void add(point &a)
	{
		for (vector<set_t>::iterator vi = mindists.begin();
		    vi != mindists.end(); ++vi) 
			vi->insert(make_pair(dist(houses[vi-mindists.begin()], a), a));
	}

	inline void remove(point &a)
	{
		for (vector<set_t>::iterator vi = mindists.begin();
		    vi != mindists.end(); ++vi)
			vi->erase(make_pair(dist(houses[vi-mindists.begin()], a), a));
	}

	void run()
	{
		double temperature = 1e10;
		int c = cost();
		fprintf(stderr, "%d\n", c);
		while (temperature > 1e-300) {
			int k = random() % bacons.size(),
			    m = random() % candidates.size();
			remove(bacons[k]);
			add(candidates[m]);
			int nc = cost();
			if (nc < c || randpos() < exp((c - nc) / temperature)) {
				swap(bacons[k], candidates[m]);
				if (nc < c)
					fprintf(stderr, "%d\n", nc);
				c = nc;
			} else {
				remove(candidates[m]);
				add(bacons[k]);
			}
			temperature *= 0.9; // you might want to change this
		}
		printf("%d\n", c);
		for (vector<point>::iterator vi = bacons.begin();
		    vi != bacons.end(); ++vi)
			printf("(%d,%d)\n", vi->first, vi->second);
		printf("%d\n", c);
	}
};

int
main(int argc, char *argv[])
{
	srandom(time(NULL));
	if (argc == 2)
		freopen(argv[1], "r", stdin);
	int w, h, b;
	scanf("%dx%d%d", &w, &h, &b);
	vector<point> p, c;
	for (int i = 0; i < h; i++) {
		static char line[1024];
		scanf("%s", line);
		for (int j = 0; j < w; j++)
			if (line[j] == 'P')
				p.push_back(make_pair(i, j));
			else
				c.push_back(make_pair(i, j));
	}
	findbacons(p, c, b).run();
	return 0;
}

File: bacongraph/saop.py
#!/usr/bin/python

# the optimized version for sa.py, using the heapq priority queue
# ./saop.py < input

import random
import math
import heapq

rand = random.Random()

def dist((a,b), (c,d)):
    return abs(a-c) + abs(b-d)

def cost(bacons, mindists):
    ret = 0
    good = set(bacons)
    for x in mindists:
        while True:
            a, b = x[0]
            if b not in good:
                heapq.heappop(x)
            else:
                ret += a
                break
    return ret

def choose(bacons, candidates):
    k = rand.randint(0, len(bacons)-1)
    m = rand.randint(0, len(candidates)-1)
    return (k, m)

def run(houses, candidates, b):
    # inits
    temperate = 10**3
    rand.shuffle(candidates)
    bacons = candidates[:b]
    candidates = candidates[b:]
    mindists = []
    for h in houses:
        s = [(dist(h, x), x) for x in bacons]
        heapq.heapify(s)
        mindists.append(s)
    c = cost(bacons, mindists)
    print c
    bestc, bestb = c, bacons[:]

    # SA iteration
    while temperate > 1e-300:
        k, m = choose(bacons, candidates)
        bacons[k], candidates[m] = candidates[m], bacons[k]
        for i, x in enumerate(mindists):
            heapq.heappush(x, (dist(houses[i], bacons[k]), bacons[k]))
        nc = cost(bacons, mindists)
        if nc < c or rand.uniform(0,1) < math.exp((c - nc) / temperate):
            if nc < bestc:
                print nc
                bestc = nc
                bestb = bacons[:]
            if nc > c:
                print math.exp((c-nc)/temperate)
            c = nc
        else: # swap back
            bacons[k], candidates[m] = candidates[m], bacons[k]
            for i, x in enumerate(mindists):
                heapq.heappush(x, (dist(houses[i], bacons[k]), bacons[k]))
        temperate *= 0.9

    return (bestc, bestb)

wh, b = raw_input().split()
width, height = map(int, wh.split('x'))
b = int(b)
houses, candidates = [], []
for x in xrange(height):
    s = raw_input()
    for y in xrange(width):
        if s[y] == 'P':
            houses.append((x,y))
        else:
            candidates.append((x,y))

print run(houses, candidates, b)
